sbncode/Calibration/TrackCaloSkimmer_module.cc

```diff
@@ -0,0 +1,460 @@
+    const std::vector<art::Ptr<recob::Hit>> &hits,
+    const std::vector<const recob::TrackHitMeta*> &thms,
+    const std::vector<art::Ptr<anab::Calorimetry>> &calo,
+    const std::map<geo::WireID, art::Ptr<raw::RawDigit>> &rawdigits,
+    const std::vector<GlobalTrackInfo> &tracks) {
+
+  // Reset the track object
+  *fTrack = sbn::TrackInfo();
+
+  // Reset other persistent info
+  fSnippetCount.clear();
+
+  // Fill top level stuff
+  fTrack->meta = fMeta;
+  fTrack->t0 = t0;
+  fTrack->id = track.ID();
+  fTrack->clear_cosmic_muon = pfp.Parent() == recob::PFParticle::kPFParticlePrimary;
+  fTrack->ndaughters = pfp.Daughters().size();
+
+  fTrack->length = track.Length();
+  fTrack->start_x = track.Start().X();
+  fTrack->start_y = track.Start().Y();
+  fTrack->start_z = track.Start().Z();
+  fTrack->end_x = track.End().X();
+  fTrack->end_y = track.End().Y();
+  fTrack->end_z = track.End().Z();
+  fTrack->dir_x = track.StartDirection().X();
+  fTrack->dir_y = track.StartDirection().Y();
+  fTrack->dir_z = track.StartDirection().Z();
+
+  // Fill each hit
+  for (unsigned i_hit = 0; i_hit < hits.size(); i_hit++) {
+    sbn::HitInfo hinfo = MakeHit(*hits[i_hit], hits[i_hit].key(), *thms[i_hit], track, calo, rawdigits);
+    if (hinfo.plane == 0) {
+      fTrack->hits0.push_back(hinfo);
+    }
+    else if (hinfo.plane == 1) {
+      fTrack->hits1.push_back(hinfo);
+    } 
+    else if (hinfo.plane == 2) {
+      fTrack->hits2.push_back(hinfo);
+    }
+  }
+
+  // Hit summary info
+  fTrack->hit_min_time_p0_tpcE = HitMinTime(fTrack->hits0, true);
+  fTrack->hit_max_time_p0_tpcE = HitMaxTime(fTrack->hits0, true);
+  fTrack->hit_min_time_p0_tpcW = HitMinTime(fTrack->hits0, false);
+  fTrack->hit_max_time_p0_tpcW = HitMaxTime(fTrack->hits0, false);
+  fTrack->hit_min_time_p1_tpcE = HitMinTime(fTrack->hits1, true);
+  fTrack->hit_max_time_p1_tpcE = HitMaxTime(fTrack->hits1, true);
+  fTrack->hit_min_time_p1_tpcW = HitMinTime(fTrack->hits1, false);
+  fTrack->hit_max_time_p1_tpcW = HitMaxTime(fTrack->hits1, false);
+  fTrack->hit_min_time_p2_tpcE = HitMinTime(fTrack->hits2, true);
+  fTrack->hit_max_time_p2_tpcE = HitMaxTime(fTrack->hits2, true);
+  fTrack->hit_min_time_p2_tpcW = HitMinTime(fTrack->hits2, false);
+  fTrack->hit_max_time_p2_tpcW = HitMaxTime(fTrack->hits2, false);
+
+  // Save information on a fit to the end of the track
+  if (fDoTailFit) DoTailFit();
+
+  // get information on nearby tracks
+  for (const GlobalTrackInfo &othr: tracks) {
+    if ((track.End() - othr.start).r() < 50. || (track.End() - othr.end).r() < 50.) {
+      fTrack->tracks_near_end_dist.push_back(std::min((track.End() - othr.start).r(), (track.End() - othr.end).r())); 
+      fTrack->tracks_near_end_costh.push_back(
+        (track.End() - othr.start).r() < (track.End() - othr.end).r() ? 
+          track.EndDirection().Dot(othr.dir) : track.EndDirection().Dot(othr.enddir)); 
+    }
+  }
+
+  // Save?
+  bool select = false;
```

Comment:
could alternatively default to true, and only not select if all of the selection tools return false.

Reply:
That's what the code currently does?

Reply:
Yeah, but it took me a bit to convince myself of that. Modifying the code may make that more clear.

Reply:
It's just De Morgan (https://en.wikipedia.org/wiki/De_Morgan%27s_laws). Also -- one thing I want to do in the piece of code is to save which tool selected the track (so as to tell a cathode crosser from a stopping track, for example). That is most easily done this way.

Reply:
OK, that's fine. Note if the goal is to know which tool passed, you stop as soon as the first tool passes, so ... you may want to consider a different way to do that.